reVX/utilities/forecasts.py

```diff
@@ -0,0 +1,386 @@
+    def bias_correct_fcst(actuals, fcsts):
+        """
+        Bias correct forecast data
+
+        Parameters
+        ----------
+        actuals : ndarray
+            Timeseries actuals (time x sites)
+        fcsts : ndarray
+            Timeseries forecats (time x sites)
+
+
+        Returns
+        -------
+        fcsts : ndarray
+            Bias corrected forecasts
+        """
+        bc_factors = actuals.sum(axis=0) / fcsts.sum(axis=0)
+        fcsts = fcsts * bc_factors
+        actuals_max = actuals.max(axis=0)
+        mask = fcsts >= actuals_max
+        fcsts = np.where(mask, actuals_max, fcsts)
+
+        return fcsts
+
+    @classmethod
+    def blend_fcsts(cls, actuals, fcsts, fcst_perc):
+        """
+        Bias correct and blend forecasts with actuals
+
+        Parameters
+        ----------
+        actuals : ndarray
+            Timeseries actuals (time x sites)
+        fcsts : ndarray
+            Timeseries forecats (time x sites)
+        fcst_perc : float
+            Percentage of forecast to use for blending
+
+        Returns
+        -------
+        fcsts : ndarray
+            Bias corrected and blended forecasts
+        """
+        fcsts = cls.bias_correct_fcst(actuals, fcsts)
+        fcsts = ((actuals * (1 - fcst_perc)) + (fcsts * fcst_perc))
+
+        return fcsts
+
+    @classmethod
+    def _correct(cls, fcst, actuals, fcst_perc=None):
+        """
+        Correct given data
+
+        Parameters
+        ----------
+        actuals : ndarray
+            Timeseries actuals (time x sites)
+        fcsts : ndarray
+            Timeseries forecats (time x sites)
+        fcst_perc : float, optional
+            Percentage of forecast to use for blending, by default None
+
+        Returns
+        -------
+        fcsts : ndarray
+            Corrected forecasts
+        """
+        mae = cls.compute_mae(actuals, fcst)
+        logger.debug('Forecast agg MAE: {:.4f}, ave MAE {:.4f}'
+                     .format(*mae))
+
+        if fcst_perc is not None:
+            logger.info('Bias correcting and blending forecasts with '
+                        '{:}% actuals'.format((1 - fcst_perc) * 100))
+            fcst = cls.blend_fcsts(actuals, fcst, fcst_perc)
+        else:
+            logger.info('Bias correcting forecasts')
+            fcst = cls.bias_correct_fcst(actuals, fcst)
+
+        mae = cls.compute_mae(actuals, fcst)
+        logger.debug('Corrected forecast agg MAE: {:.4f}, '
+                     'ave MAE {:.4f}'.format(*mae))
+
+        return fcst
+
+    @staticmethod
+    def compute_mae(actuals, fcsts):
+        """
+        Compute aggregate and average MAE between actuals and forecasts
+
+        Parameters
+        ----------
+        actuals : ndarray
+            Timeseries actuals (time x sites)
+        fcsts : ndarray
+            Timeseries forecats (time x sites)
+
+        Returns
+        -------
+        agg_mae : float
+            Aggregate MAE for all sites
```

Comment:
What is this? The sum of the MAE for all sites?

Reply:
As I remember yes. Look at the over all MAE for all sites vs the average MAE across sites